lib/Sema/CSGen.cpp

```diff
@@ -3543,17 +3543,38 @@ namespace {
     
     Type visitUnresolvedPatternExpr(UnresolvedPatternExpr *expr) {
       // Encountering an UnresolvedPatternExpr here means we have an invalid
-      // ExprPattern with a Pattern node like 'let x' nested in it. Record a
-      // fix, and assign ErrorTypes to any VarDecls bound.
+      // ExprPattern with a Pattern node like 'let x' nested in it.
+      // Assign ErrorTypes to any VarDecls bound.
       auto *locator = CS.getConstraintLocator(expr);
       auto *P = expr->getSubPattern();
-      CS.recordFix(IgnoreInvalidPatternInExpr::create(CS, P, locator));
-
       P->forEachVariable([&](VarDecl *VD) {
         CS.setType(VD, ErrorType::get(CS.getASTContext()));
       });
-
-      return CS.createTypeVariable(locator, TVO_CanBindToHole);
+      auto *typeVariable = CS.createTypeVariable(locator, TVO_CanBindToHole);
+      // Record the more general fix.
+      auto ignoreInvalidPattern = [&]() {
+        CS.recordFix(IgnoreInvalidPatternInExpr::create(CS, P, locator));
+        return typeVariable;
+      };
+      // Check that we are in a CheckedCastExpr. Also, make sure we are
+      // in an argument of `~=` that is _not_ inside a CallExpr;
+      // this avoids cases like: `case let f(x as? Int) = y {}`,
+      // where the fix is less useful.
+      auto *E = CS.getParentExpr(expr);
+      while (E && !isa<CheckedCastExpr>(E))
+        E = CS.getParentExpr(E);
+      auto *castExpr = cast_or_null<CheckedCastExpr>(E);
+      if (!castExpr)
+        return ignoreInvalidPattern();
+      auto *parent = CS.getParentExpr(castExpr);
+      while (parent && !isa<BinaryExpr>(parent))
+        parent = isa<CallExpr>(parent) ? nullptr : CS.getParentExpr(parent);
+      auto *BE = cast_or_null<BinaryExpr>(parent);
+      if (!BE || !isPatternMatchingOperator(BE->getFn()))
+        return ignoreInvalidPattern();
+      CS.recordFix(IgnoreInvalidCastInPattern::create(
+          CS, CS.getConstraintLocator(castExpr)));
```

Comment:
This looks like it belongs in `InvalidPatternInExprFailure`, I don't think we need a special fix kind for this, only a tailored diagnostic type.

Reply:
Yeah I like that better too.